sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala

```diff
@@ -0,0 +1,285 @@
+        val splitIx = (tbl.getRowCount / 2).toInt
+        withResource(tbl.contiguousSplit(splitIx)) { cts =>
+          val batches = cts.map(_.getTable).map(GpuColumnVector.from(_, spillable.dataTypes))
+          val spillables = batches.map { b =>
+            SpillableColumnarBatch(
+              b,
+              SpillPriorities.ACTIVE_BATCHING_PRIORITY,
+              // TODO: need to figure out how to pick the right callback
+              RapidsBuffer.defaultSpillCallback)
+          }
+          require(spillables.length == 2)
+          (spillables.head, spillables.last)
+        }
+      }
+    }
+    Seq(firstHalf, secondHalf)
+  }
+
+  private def toSpillableIterator(
+      it: Iterator[ColumnarBatch]): Iterator[SpillableColumnarBatch] = {
+    new Iterator[SpillableColumnarBatch] {
+      override def hasNext: Boolean = it.hasNext
+      override def next(): SpillableColumnarBatch = {
+        SpillableColumnarBatch(
+          it.next(),
+          SpillPriorities.ACTIVE_ON_DECK_PRIORITY,
+          RapidsBuffer.defaultSpillCallback)
+      }
+    }
+  }
+
+  def withRetry(
+      input: SpillableColumnarBatch)
+      (fn: ColumnarBatch => ColumnarBatch): Iterator[SpillableColumnarBatch] = {
+    toSpillableIterator(new RmmRapidsRetryIterator(Seq(input).iterator, fn))
+  }
+
+  def withRetry(
+      input: SpillableColumnarBatch,
+      splitPolicy: SpillableColumnarBatch => Seq[SpillableColumnarBatch])
+      (fn: ColumnarBatch => ColumnarBatch): Iterator[SpillableColumnarBatch] = {
+    toSpillableIterator(new RmmRapidsRetryIterator(Seq(input).iterator, fn, splitPolicy))
+  }
+
+  def withRetry(
+      input: SpillableColumnarBatch,
+      splitPolicy: SpillableColumnarBatch => Seq[SpillableColumnarBatch],
+      mergePolicy: Seq[SpillableColumnarBatch] => SpillableColumnarBatch)
+      (fn: ColumnarBatch => ColumnarBatch): Iterator[SpillableColumnarBatch] = {
+    toSpillableIterator(new RmmRapidsRetryIterator(
+      Seq(input).iterator,
+      fn,
+      splitPolicy,
+      mergePolicy))
+  }
+}
+
+/**
+ * RmmRapidsRetryHelper is used to attempt work on the GPU with the ability to retry it,
+ * abstracting the retry logic from the calling code.
+ *
+ * @param input a SpillableColumnarBatch that we would like to attempt work with
+ * @param splitPolicy a function that can split a SpillableColumnarBatch into multiple
+ *                    spillable batches, can be set to null in case splits are not handled.
+ * @param mergePolicy a function that can be used to merge SpillableColumnarBatches that had
+ *                    been previously split, can be set to null in case merging is not handled
+ *                    or not required.
+ */
+class RmmRapidsRetryIterator[T](
+    input: Iterator[SpillableColumnarBatch],
+    fn: ColumnarBatch => T,
+    splitPolicy: SpillableColumnarBatch => Seq[SpillableColumnarBatch],
+    mergePolicy: Seq[SpillableColumnarBatch] => SpillableColumnarBatch)
```

Comment:
It looks like merge policy does not work so I am just going to stop reviewing this more until you give the the heads up @abellina